src/transport/manager/peer_state.rs

```diff
@@ -0,0 +1,928 @@
+                        };
+                    }
+                    _ => (),
+                }
+            }
+            _ => (),
+        }
+
+        false
+    }
+
+    /// Returns `true` if the last transport failed to open.
+    pub fn on_open_failure(&mut self, transport: SupportedTransport) -> bool {
+        match self {
+            Self::Opening { transports, .. } => {
+                transports.remove(&transport);
+
+                if transports.is_empty() {
+                    *self = Self::Disconnected { dial_record: None };
+                    return true;
+                }
+
+                return false;
+            }
+            _ => false,
+        }
+    }
+
+    /// Returns `true` if the connection was opened.
+    pub fn on_connection_opened(&mut self, record: ConnectionRecord) -> bool {
+        match self {
+            Self::Opening { .. } => {
+                // TODO: Litep2p did not check previously if the
+                // connection record is valid or not, in terms of having
+                // the same connection ID and the address part of the
+                // address set.
+
+                *self = Self::Dialing {
+                    dial_record: record.clone(),
+                };
+
+                true
+            }
+            _ => false,
+        }
+    }
+}
+
+/// The connection record keeps track of the connection ID and the address of the connection.
+///
+/// The connection ID is used to track the connection in the transport layer.
+/// While the address is used to keep a healthy view of the network for dialing purposes.
+///
+/// # Note
+///
+/// The structure is used to keep track of:
+///
+///  - dialing state for outbound connections.
+///  - established outbound connections via [`PeerState::Connected`].
+///  - established inbound connections via `PeerContext::secondary_connection`.
+#[derive(Debug, Clone, Hash, PartialEq)]
+pub struct ConnectionRecord {
+    /// Address of the connection.
+    ///
+    /// The address must contain the peer ID extension `/p2p/<peer_id>`.
+    pub address: Multiaddr,
+
+    /// Connection ID resulted from dialing.
+    pub connection_id: ConnectionId,
+}
+
+impl ConnectionRecord {
+    /// Construct a new connection record.
+    pub fn new(peer: PeerId, address: Multiaddr, connection_id: ConnectionId) -> Self {
+        Self {
+            address: Self::ensure_peer_id(peer, address),
+            connection_id,
+        }
+    }
+
+    /// Create a new connection record from the peer ID and the endpoint.
+    pub fn from_endpoint(peer: PeerId, endpoint: &Endpoint) -> Self {
+        Self {
+            address: Self::ensure_peer_id(peer, endpoint.address().clone()),
+            connection_id: endpoint.connection_id(),
+        }
+    }
+
+    /// Ensures the peer ID is present in the address.
+    fn ensure_peer_id(peer: PeerId, address: Multiaddr) -> Multiaddr {
+        if !std::matches!(address.iter().last(), Some(Protocol::P2p(_))) {
```

Comment:
May be overwrite the peer ID and add a warning in case of mismatch to catch potential problems?